### PR TITLE
[SPARK-10174][SQL] refactor out project, filter, ordering generator from SparkPlan

### PR DESCRIPTION
**Pull request by cloud-fan**
These methods don't depend on `SparkPlan`'s states and sometimes we need to create these operators outside of `SparkPlan`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CommonOperatorGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{GenerateOrdering, GeneratePredicate, GenerateMutableProjection, GenerateProjection}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.types.DataType
+
+object CommonOperatorGenerator extends Logging {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CommonOperatorGenerator.scala]: What about making this a `trait` and making `codegenEnabled` an abstract method?

**Comment by cloud-fan** [comment]
Hi @marmbrus , I think making `CommonOperatorGenerator` object is better than trait, as it simplifies the class inheritance and helps us get rid of serialization problems. What do you think?

**Comment by marmbrus** [comment]
What is complicated about inheritance or serialization for a trait that only defines methods?  This is just sugar to make the invocation shorter and a common pattern throughout the code base.

**Comment by cloud-fan** [comment]
If we call generator method like `newProject` inside RDD function closure, then we will have to serialize the outer class which mix-in `CommonOperatorGenerator`, and it's unnecessary. However, if we make `CommonOperatorGenerator` an object, then we only need to serialize the generator object not the whole outer class.

**Comment by marmbrus** [comment]
Mixing it in isn't going to change the size of the class itself, and in nearly all cases aren't we pulling in the outer class already?

```
scala> class A extends Serializable { val x = 1 }
defined class A

scala> org.apache.spark.util.Utils.serialize(new A).length
res3: Int = 140

scala> trait T { def y = 2 }
defined trait T

scala> class B extends T with Serializable { val x = 1 }
defined class B

scala> org.apache.spark.util.Utils.serialize(new B).length
res4: Int = 140
```

**Comment by cloud-fan** [comment]
An example is `HadoopFsRelation`, it needs to build a mutable projection inside `RDD.mapPartitions` but itself is not serializable. Anyway we can take it as a special case and still build mutable projection manually instead of mix-in `CommonOperatorGenerator` if you prefer the trait solution.

**Comment by andrewor14** [comment]
@cloud-fan this patch has mostly gone stale and I think given the changes in master it would take significant effort to resolve the conflicts. Can you close this patch for now?

**Pull request closed without merging**